bumble/device.py

```diff
@@ -2778,84 +2778,60 @@ def on_authentication_user_confirmation_request(self, connection, code):
         pairing_config = self.pairing_config_factory(connection)
         io_capability = pairing_config.delegate.classic_io_capability
 
-        # Respond
+        # See Bluetooth spec @ Vol 3, Part C 5.2.2.6
         if io_capability == HCI_DISPLAY_YES_NO_IO_CAPABILITY:
-            if connection.peer_pairing_io_capability in (
-                HCI_DISPLAY_YES_NO_IO_CAPABILITY,
-                HCI_DISPLAY_ONLY_IO_CAPABILITY,
-            ):
+
+            async def prompt() -> bool:
                 # Display the code and ask the user to compare
-                async def prompt():
-                    return (
-                        await pairing_config.delegate.compare_numbers(code, digits=6),
-                    )
+                return await pairing_config.delegate.compare_numbers(code, digits=6)
 
-            else:
-                # Ask the user to confirm the pairing, without showing a code
-                async def prompt():
-                    return await pairing_config.delegate.confirm()
+        else:
 
-            async def confirm():
-                if await prompt():
-                    await self.host.send_command(
-                        HCI_User_Confirmation_Request_Reply_Command(
-                            bd_addr=connection.peer_address
-                        )
+            async def prompt() -> bool:
+                # Ask the user to confirm the pairing, without display
+                return await pairing_config.delegate.confirm()
+
+        async def confirm() -> None:
+            if await connection.abort_on('disconnection', prompt()):
+                await self.host.send_command(
+                    HCI_User_Confirmation_Request_Reply_Command(  # type: ignore[call-arg]
+                        bd_addr=connection.peer_address
                     )
-                else:
-                    await self.host.send_command(
-                        HCI_User_Confirmation_Request_Negative_Reply_Command(
-                            bd_addr=connection.peer_address
-                        )
+                )
+            else:
+                await self.host.send_command(
+                    HCI_User_Confirmation_Request_Negative_Reply_Command(  # type: ignore[call-arg]
+                        bd_addr=connection.peer_address
                     )
+                )
 
-            AsyncRunner.spawn(connection.abort_on('disconnection', confirm()))
-            return
-
-        if io_capability == HCI_DISPLAY_ONLY_IO_CAPABILITY:
-            # Display the code to the user
-            AsyncRunner.spawn(pairing_config.delegate.display_number(code, 6))
-
-        # Automatic confirmation
```

Comment:
I don't think we can remove this automatic confirmation without prompt. When the responder has no input (`DisplayOnly`), we can't ask the user to confirm. The spec does say "with automatic confirmation".

Reply:
I think it makes sense to call the delegate even for auto-confirm cases. Maybe the way to do it is to add a `auto=False` parameter to `confirm()` so that a "real" delegate (i.e with a user) can know that it's being called but isn't supposed to actually prompt the user. For auto-confirm flows, we'd call `confirm(auto=True)`, and just `confirm()` in the non-auto confirm cases.